### PR TITLE
Account api merge

### PR DESCRIPTION
**Pull request by TrueFlame42**

### CODE REVIEW

File: aspnet/RVTR.Account.DataContext/Repositories/AccountRepository.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using RVTR.Account.ObjectModel.Models;
+namespace RVTR.Account.DataContext.Repositories
+{
+  /// <summary>
+  /// Represents the _Repository_ generic
+  /// </summary>
+  /// <typeparam name="TEntity"></typeparam>
+  public class AccountRepository : Repository<AccountModel>
```
Comment (escalonn) [inline aspnet/RVTR.Account.DataContext/Repositories/AccountRepository.cs:12]: wrong/bad documentation on this class

File: aspnet/RVTR.Account.DataContext/Repositories/Repository.cs [hunk 2]
```diff
@@ -27,4 +27,4 @@ public Repository(AccountContext context)
 
     public virtual void Update(TEntity entry) => _db.Update(entry);
   }
-}
+}
```
Comment (escalonn) [inline aspnet/RVTR.Account.DataContext/Repositories/Repository.cs:30]: several bad formatting diffs like this remain

File: aspnet/RVTR.Account.DataContext/SeedDb.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
+﻿using RVTR.Account.ObjectModel.Models;
+using System.Collections.Generic;
+using System;
+using System.Linq;
+
+namespace RVTR.Account.DataContext
+{
+  /// <summary>
+  /// Represents the _Account_ context
+  /// </summary>
```
Comment (escalonn) [inline aspnet/RVTR.Account.DataContext/SeedDb.cs:10]: why are these comments here, they're blatantly wrong

File: aspnet/RVTR.Account.DataContext/SeedDb.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  public static class SeedDb
+  {
+    public static void Seed(AccountContext ctx)
```
Comment (escalonn) [inline aspnet/RVTR.Account.DataContext/SeedDb.cs:13]: no reason for this to not be async

File: aspnet/RVTR.Account.DataContext/SeedDb.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+          City = "City",
+          Country = "Country",
+          PostalCode = "21345",
+          StateProvince = "Somewhere",
+          Street = "123 elm street",
+          AccountId = -1,
+        },
+        Name = "Name",
+        Payments = new List<PaymentModel>()
+                            {
```
Comment (escalonn) [inline aspnet/RVTR.Account.DataContext/SeedDb.cs:35]: this indentation is all over the place

File: aspnet/RVTR.Account.ObjectModel/Models/AccountModel.cs [hunk 6]
```diff
@@ -8,11 +10,25 @@ namespace RVTR.Account.ObjectModel.Models
   /// </summary>
   public class AccountModel : IValidatableObject
   {
+    [Key]
+    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
```
Comment (escalonn) [inline aspnet/RVTR.Account.ObjectModel/Models/AccountModel.cs:14]: why are these here, they don't belong
same for every other class

File: aspnet/RVTR.Account.ObjectModel/Models/AccountModel.cs [hunk 7]
```diff
@@ -8,11 +10,25 @@ namespace RVTR.Account.ObjectModel.Models
...
+      get => _name;
+      set
+      {
+        if (string.IsNullOrEmpty(value))
+        {
+          throw new ArgumentException("Account name cannot be null.", nameof(value));
+        }
+        _name = value;
+      }
+    }
```
Comment (escalonn) [inline aspnet/RVTR.Account.ObjectModel/Models/AccountModel.cs:31]: the validate method below is really where this is meant to be, i think

File: aspnet/RVTR.Account.ObjectModel/Models/AddressModel.cs [hunk 8]
```diff
@@ -8,17 +10,91 @@ namespace RVTR.Account.ObjectModel.Models
...
+        {
+          throw new ArgumentException("Street cannot be null.", nameof(value));
+        }
+        _street = value;
+      }
+    }
+
+    [ForeignKey("Account")]
+    [Required]
+    public int? AccountId { get; set; }
```
Comment (escalonn) [inline aspnet/RVTR.Account.ObjectModel/Models/AddressModel.cs:94]: what? nullable but then Required? what is the sense in that?
for that matter what is the purpose of any of the changes to this file?

File: aspnet/RVTR.Account.ObjectModel/Models/PaymentModel.cs [hunk 9]
```diff
@@ -8,11 +10,47 @@ namespace RVTR.Account.ObjectModel.Models
   /// </summary>
   public class PaymentModel : IValidatableObject
   {
+    [Key]
+    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
     public int Id { get; set; }
 
-    public BankCardModel BankCard { get; set; }
+    public DateTime cardExpirationDate { get; set; }
```
Comment (escalonn) [inline aspnet/RVTR.Account.ObjectModel/Models/PaymentModel.cs:17]: naming convention...

File: aspnet/RVTR.Account.UnitTesting/Tests/AccountControllerTest.cs [hunk 10]
```diff
@@ -79,5 +80,15 @@ public async void Test_Controller_Put()
 
       Assert.NotNull(resultPass);
     }
+    //expect error if account isn't found
+    [Fact]
+    public async void Test_404_Response()
+    {
+      var result = await _controller.Get(-100);
+
+      Assert.IsType<Microsoft.AspNetCore.Mvc.NotFoundObjectResult>(result);
```
Comment (escalonn) [inline aspnet/RVTR.Account.UnitTesting/Tests/AccountControllerTest.cs:89]: just use a using directive, why put the full name
bad whitespace in this file

File: aspnet/RVTR.Account.WebApi/Controllers/AccountController.cs [hunk 11]
```diff
@@ -31,80 +34,153 @@ public AccountController(ILogger<AccountController> logger, UnitOfWork unitOfWor
...
+          _logger.LogDebug("Deleting an account by its ID number...");
+        }
         await _unitOfWork.Account.DeleteAsync(id);
         await _unitOfWork.CommitAsync();
 
-        return Ok();
+        if (_logger != null)
+        {
+          _logger.LogInformation($"Deleted the account.");
+        }
```
Comment (escalonn) [inline aspnet/RVTR.Account.WebApi/Controllers/AccountController.cs:58]: we'd want the ID logged for this sort of thing

File: aspnet/RVTR.Account.WebApi/Controllers/AccountController.cs [hunk 12]
```diff
@@ -31,80 +34,153 @@ public AccountController(ILogger<AccountController> logger, UnitOfWork unitOfWor
...
       await _unitOfWork.Account.InsertAsync(account);
       await _unitOfWork.CommitAsync();
 
-      return Accepted(account);
+      if (_logger != null)
+      {
+        _logger.LogInformation($"Successfully added the account {account}.");
+      }
+      return Ok(MessageObject.Success);
+
```
Comment (escalonn) [inline aspnet/RVTR.Account.WebApi/Controllers/AccountController.cs:144]: 200 OK with some success message is not the expected standard for a POST that creates something; if the POST created something, then we need a 201 with a location header and a representation of the resource

File: aspnet/RVTR.Account.WebApi/Controllers/AccountController.cs [hunk 13]
```diff
@@ -31,80 +34,153 @@ public AccountController(ILogger<AccountController> logger, UnitOfWork unitOfWor
...
 
     /// <summary>
-    ///
+    /// Add an account 
     /// </summary>
     /// <param name="account"></param>
     /// <returns></returns>
     [HttpPost]
-    public async Task<IActionResult> Post(AccountModel account)
+    [ProducesResponseType(typeof(AccountModel), StatusCodes.Status200OK)]
```
Comment (escalonn) [inline aspnet/RVTR.Account.WebApi/Controllers/AccountController.cs:128]: these documentation attributes are wrong. you are not returning an accountmodel in the body, you're returning some responseobject thing

File: aspnet/RVTR.Account.WebApi/Controllers/AddressController.cs [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+        if (_logger != null)
+        {
+          _logger.LogInformation($"Retrieved the address with ID: {id}.");
+        }
+        return Ok(theAddress);
+      }
+      if (_logger != null)
+      {
+        _logger.LogWarning($"Address with ID number {id} does not exist.");
+      }
```
Comment (escalonn) [inline aspnet/RVTR.Account.WebApi/Controllers/AddressController.cs:118]: this way of logging is overly in the way of the rest of the code. we have `?.` operator for this use case, but itd also be fine to just assume the logger isnt null or throw and exception in ctor if it is

File: aspnet/RVTR.Account.WebApi/RVTR.Account.WebApi.csproj [hunk 15]
```diff
@@ -11,17 +11,24 @@
...
   <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="3.1.7" />
     <PackageReference Include="microsoft.aspnetcore.mvc.versioning" Version="4.1.1" />
     <PackageReference Include="microsoft.aspnetcore.mvc.versioning.apiexplorer" Version="4.1.1" />
     <PackageReference Include="npgsql.entityframeworkcore.postgresql" Version="3.1.3" />
     <PackageReference Include="swashbuckle.aspnetcore" Version="5.4.1" />
     <PackageReference Include="zipkin4net" Version="1.5.0" />
     <PackageReference Include="zipkin4net.middleware.aspnetcore" Version="1.5.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
+    <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="3.1.7" />
```
Comment (escalonn) [inline aspnet/RVTR.Account.WebApi/RVTR.Account.WebApi.csproj:22]: why is the newtonsoft one in here twice

File: aspnet/RVTR.Account.WebApi/RVTR.Account.WebApi.csproj [hunk 16]
```diff
@@ -11,17 +11,24 @@
...
   </ItemGroup>
 
   <ItemGroup>
     <ProjectReference Include="..\RVTR.Account.ObjectModel\RVTR.Account.ObjectModel.csproj" />
     <ProjectReference Include="..\RVTR.Account.DataContext\RVTR.Account.DataContext.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="ResponseObjects\" />
+  </ItemGroup>
```
Comment (escalonn) [inline aspnet/RVTR.Account.WebApi/RVTR.Account.WebApi.csproj:33]: why is this here

File: aspnet/RVTR.Account.WebApi/Startup.cs [hunk 17]
```diff
@@ -72,20 +77,24 @@ public void ConfigureServices(IServiceCollection services)
...
+      loggerFactory.AddFile("Logs/ts-{Date}.txt");
+
       if (hostEnvironment.IsDevelopment())
       {
         applicationBuilder.UseDeveloperExceptionPage();
       }
 
       applicationBuilder.UseZipkin();
       applicationBuilder.UseTracing("accountapi.rest");
+      applicationBuilder.UseHttpsRedirection();
```
Comment (escalonn) [inline aspnet/RVTR.Account.WebApi/Startup.cs:97]: why has this been added?

File: aspnet/RVTR.Account.ObjectModel/Models/AccountModel.cs (hunk 6 above)
Comment (pamelasoulis) [reply aspnet/RVTR.Account.ObjectModel/Models/AccountModel.cs:14]: Addressed this in commit #f8bd653

File: aspnet/RVTR.Account.ObjectModel/Models/AddressModel.cs (hunk 8 above)
Comment (pamelasoulis) [reply aspnet/RVTR.Account.ObjectModel/Models/AddressModel.cs:94]: addressed in commit f8bd653

File: aspnet/RVTR.Account.UnitTesting/Tests/AccountControllerTest.cs (hunk 10 above)
Comment (pamelasoulis) [reply aspnet/RVTR.Account.UnitTesting/Tests/AccountControllerTest.cs:89]: addressed in commit f8bd653

File: aspnet/RVTR.Account.WebApi/Controllers/AccountController.cs (hunk 11 above)
Comment (pamelasoulis) [reply aspnet/RVTR.Account.WebApi/Controllers/AccountController.cs]: addressed in commit f8bd653

File: aspnet/RVTR.Account.WebApi/Controllers/AccountController.cs (hunk 13 above)
Comment (pamelasoulis) [reply aspnet/RVTR.Account.WebApi/Controllers/AccountController.cs]: addressed in commit f8bd653

File: aspnet/RVTR.Account.WebApi/Controllers/AddressController.cs (hunk 14 above)
Comment (pamelasoulis) [reply aspnet/RVTR.Account.WebApi/Controllers/AddressController.cs]: addressed in commit f8bd653

File: aspnet/RVTR.Account.DataContext/SeedDb.cs (hunk 3 above)
Comment (pamelasoulis) [reply aspnet/RVTR.Account.DataContext/SeedDb.cs:11]: addressed in commit dd6eb70

File: aspnet/RVTR.Account.WebApi/Controllers/AccountController.cs (hunk 12 above)
Comment (pamelasoulis) [reply aspnet/RVTR.Account.WebApi/Controllers/AccountController.cs]: addressed this comment in dd6eb70

File: aspnet/RVTR.Account.WebApi/Controllers/AccountController.cs [hunk 18]
```diff
@@ -125,23 +113,19 @@ public async Task<IActionResult> Get(int id)
...
 
-      if (_logger != null)
-      {
-        _logger.LogInformation($"Successfully added the account {account}.");
-      }
-      return Ok(MessageObject.Success);
-
+      _logger.LogInformation($"Successfully added the account {account}.");
+      
+      return Accepted(account);
```
Comment (escalonn) [inline aspnet/RVTR.Account.WebApi/Controllers/AccountController.cs:127]: like i said it's 201 for creation, not 202.
